apps/files/src/store/renaming.ts

```diff
@@ -8,11 +8,59 @@ import type { RenamingStore } from '../types'
 import axios, { isAxiosError } from '@nextcloud/axios'
 import { emit, subscribe } from '@nextcloud/event-bus'
 import { NodeStatus } from '@nextcloud/files'
+import { DialogBuilder } from '@nextcloud/dialogs'
 import { t } from '@nextcloud/l10n'
-import { basename, dirname } from 'path'
+import { basename, dirname, extname } from 'path'
 import { defineStore } from 'pinia'
 import logger from '../logger'
 import Vue from 'vue'
+import IconCancel from '@mdi/svg/svg/cancel.svg?raw'
+import IconCheck from '@mdi/svg/svg/check.svg?raw'
+
+let isDialogVisible = false
+
+const showWarningDialog = (oldExtension: string, newExtension: string): Promise<boolean> => {
+	if (isDialogVisible) {
+		return Promise.resolve(false)
+	}
+
+	isDialogVisible = true
+
+	return new Promise((resolve) => {
+		const dialog = new DialogBuilder()
+			.setName(t('files', 'Extension Change Warning'))
+			.setText(t(
+				'files',
+				'You are attempting to change the file extension from "{old}" to "{new}". This may affect how the file is handled. Do you want to proceed?',
+				{ old: oldExtension || t('files', 'none'), new: newExtension || t('files', 'none') },
+			))
+			.setButtons([
+				{
+					label: t('files', 'Cancel'),
+					icon: IconCancel,
+					type: 'secondary',
+					callback: () => {
+						isDialogVisible = false
+						resolve(false)
+					},
+				},
+				{
+					label: t('files', 'Proceed'),
+					icon: IconCheck,
+					type: 'primary',
+					callback: () => {
+						isDialogVisible = false
+						resolve(true)
+					},
+				},
+			])
+			.build()
+
+		dialog.show().then(() => {
+			dialog.hide()
+		})
+	})
+}
```

Comment:
Regardless of wording I think that this is a bit too nested, just make it async instead and you can reduce the nesting:

```suggestion
async function showWarningDialog(oldExtension: string, newExtension: string): Promise<boolean> {
	if (isDialogVisible) {
		return false
	}

	let response = false
	isDialogVisible = true
	const dialog = new DialogBuilder()
		.setName(t('files', 'Extension Change Warning'))
		.setText(t(
			'files',
			'You are attempting to change the file extension from "{old}" to "{new}". This may affect how the file is handled. Do you want to proceed?',
			{ old: oldExtension || t('files', 'none'), new: newExtension || t('files', 'none') },
		))
		.setButtons([
			{
				label: t('files', 'Cancel'),
				icon: IconCancel,
				type: 'secondary',
				callback: () => {
					response = false
				},
			},
			{
				label: t('files', 'Proceed'),
				icon: IconCheck,
				type: 'primary',
				callback: () => {
					response = true
				},
			},
		])
		.build()

	try {
		await dialog.show()
		return response
	} catch (error) {
		// dialog closed without response
		return false
	} finally {
		dialogVisible = false
	}
}
```

Reply:
BTW this also fixes the issue that there is a uncaught exception when closing the dialog with your code.

(Also this allows use to easier simplify this after nc-dialogs got the latest nc-vue changes which allow to return from button callbacks and get the returned value back in `await dialog.show()`)

Reply:
Makes sense, thanks...